src/Tools/ExternalAccess/Razor/Cohost/RazorCohostLanguageClient.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Implementation.LanguageClient;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Utilities;
+using Newtonsoft.Json;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+
+/// <summary>
+/// A language server that handles requests .razor and .cshtml files. Endpoints and required services are supplied
+/// by the Razor tooling team in the Razor tooling repo.
+/// </summary>
+[ContentType(Constants.RazorLSPContentType)]
+[Export(typeof(ILanguageClient))]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class RazorCohostLanguageClient(
+    RazorLspServiceProvider lspServiceProvider,
+    IGlobalOptionService globalOptions,
+    IThreadingContext threadingContext,
+    ILspServiceLoggerFactory lspLoggerFactory,
+    ExportProvider exportProvider,
+    [Import(AllowDefault = true)] IRazorCohostCapabilitiesProvider? razorCapabilitiesProvider = null,
+    [Import(AllowDefault = true)] IRazorCustomMessageTarget? razorCustomMessageTarget = null,
+    [Import(AllowDefault = true)] IRazorCohostLanguageClientActivationService? razorCohostLanguageClientActivationService = null)
+    : AbstractInProcLanguageClient(lspServiceProvider, globalOptions, lspLoggerFactory, threadingContext, exportProvider, middleLayer: null)
+{
+    private readonly IRazorCohostLanguageClientActivationService? _razorCohostLanguageClientActivationService = razorCohostLanguageClientActivationService;
+    private readonly IRazorCohostCapabilitiesProvider? _razorCapabilitiesProvider = razorCapabilitiesProvider;
+    private readonly IRazorCustomMessageTarget? _razorCustomMessageTarget = razorCustomMessageTarget;
+
+    protected override ImmutableArray<string> SupportedLanguages => Constants.RazorLanguage;
+
+    public override object? CustomMessageTarget => _razorCustomMessageTarget;
+
+    public override Task<Connection?> ActivateAsync(CancellationToken cancellationToken)
+    {
+        if (_razorCohostLanguageClientActivationService?.ShouldActivateCohostServer() == true)
+        {
+            return base.ActivateAsync(cancellationToken);
+        }
+
+        return Task.FromResult<Connection?>(null);
```

Comment:
Sadly, this doesn't work. The platform helpfully checks this for null and just immediately throws an exception 🤦‍♂️

https://dev.azure.com/devdiv/DevDiv/_git/VSLanguageServerClient?path=/src/product/RemoteLanguage/Impl/Infrastructure/MefClientProvider/RemoteLanguageClientInstance.cs&version=GBdevelop&line=528&lineEnd=529&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

So options are ClientName, or just return no capabilities. Thoughts @dibarbet ?

I'm guessing ClientName, since activating a server to nothing isn't great, but there is a slight downside with that, that we'd be unable to do progressive moves to this server, it would be all or nothing. I don't know how important that will be though.

Reply:
So I'm not super opposed to having a server with no capabilities.  We'll just have to make sure that the server isn't trying to do anything (e.g. sending refresh notifications).

As you mention it would allow you to migrate features one at a time which I think is fairly valuable.